OneCommunityGlobal/HGNRest
# Diego enforcing eslint and prettier on staged files

<img width="757" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/120c6ba0-675c-4be8-9dce-2be0eea9148f">

# Description
Adds eslint and prettier to format the code style and enforce finding programming errors and fixing errors programatically on staged files before code is committed. While we cannot control how each developer sets up their development environment, we can enforce running prettier and eslint on staged files so only the files that are about to be committed will be inspected and formatted without having to manually perform any action. 

## Related PRS:
N/A

## Main changes explained:
- Functionality binds to git hooks with husky (pre-commit) to run eslint and prettier on staged files
- Lints staged files (not the entire codebase)
- Lints and applies code style formatting sequentially before code is committed
- Formats code style with prettier even when eslint doesn't find any linting errors or warnings.

## How to test:
1. find "Preferences: Open User Setting (JSON)" options by entering `cmd + shift + p` in vs code
![Screen Shot 2024-03-20 at 11.07.34 AM](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/6a34fc57-68ae-469f-b5be-04ce7d8e0dc9)
2. comment all of the properties and values if the JSON file (this will disable prettier from formatting code when a file is saved):
![Screen Shot 2024-03-20 at 11.09.13 AM](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/0a71d34f-cc3c-45e7-aade-63b2b1b0c069)
3. check into current branch
4. do `npm install`
5. checkout into a new branch of your own to test this functionality (since we will create a new commit to test this). eg. `git checkout -b yourname/test_eslint_prettier`
6. if running on Mac or Linux run: npm run prepare (this will make husky files executable)
7. update any file: eg. `server.js`:
![Screen Shot 2024-03-20 at 11.14.06 AM](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/f5d17d1d-d4f3-40b6-8c5a-f3894d881c97)
8. try breaking any of the `.prettierrc` defined rules
9. save the changes, add them to the staging area (git add filename), commit the changes.
10. lint-staged and prettier should run and stop the execution if there are any programming errors that eslint cannot fixed (for instance a variable that is not initialized like in the previous image); otherwise, the files should be formatted with prettier (based on the `.prettierrc` file) and fixed with eslint if any rules broken were fixed by eslint.
11. try changing another file. eg. config.js
![Screen Shot](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/ff0568dc-bf39-48ff-8225-d66bf5c573e7)
12. add only one of the files to the staging area, commit the changes. If the commit succeeds, only the staged file will be linted and formatted. The rest of the files that are not in the staging area will remain untouched.

## Screenshots or videos of changes:
### Running
https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/f364762d-cf56-41e2-b881-98eb2347b54c

## Note:
This change integrates eslint and prettier in a way that no conflicts will arise between them. Rules defined in .prettierrc will override eslint code style formatting rules (which is recommended). Eslint will only be used to catch programming errors that can lead to bugs in production.
